pkg/quota/admission/podlimitrequest/admission.go

```diff
@@ -0,0 +1,125 @@
+package podlimitrequest
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/openshift/origin/pkg/cmd/server/api"
+	"k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	"speter.net/go/exp/math/dec/inf"
+)
+
+const (
+	annEnabled         = "openshift.io/PodLimitRequestEnabled"
+	cpuBaseScaleFactor = 1000.0 / (1024.0 * 1024.0 * 1024.0) // 1000 milliCores per 1GiB
+)
+
+func init() {
+	admission.RegisterPlugin("PodLimitRequest", func(client kclient.Interface, config io.Reader) (admission.Interface, error) {
+		return newPodLimitRequest(client, config)
+	})
+}
+
+type podLimitRequest struct {
+	*admission.Handler
+	Config     api.PodLimitRequestConfig
+	KubeClient kclient.Interface
+}
+
+// newPodLimitRequest returns an admission controller for containers that
+// configurably overrides container resource request/limits
+func newPodLimitRequest(client kclient.Interface, config io.Reader) (admission.Interface, error) {
+	glog.V(5).Infof("PodLimitRequest admission controller is loaded")
+	parsed, err := ReadConfig(config)
+	glog.V(5).Infof("PodLimitRequest admission controller got config: %v\nerror: %v", parsed, err)
+	return &podLimitRequest{
+		Handler:    admission.NewHandler(admission.Create),
+		Config:     parsed,
+		KubeClient: client,
+	}, err
+}
+
+func ReadConfig(configFile io.Reader) (api.PodLimitRequestConfig, error) {
+	glog.V(5).Infof("PodLimitRequest about to read config:\n%v", configFile)
+	config := api.PodLimitRequestConfig{}
+	buffer := new(bytes.Buffer)
+	buffer.ReadFrom(configFile)
+	glog.V(5).Infof("PodLimitRequest read config:\n%v", buffer.Bytes())
+	err := yaml.Unmarshal(buffer.Bytes(), &config)
+	if err == nil && config.Enabled {
+		if config.LimitCPUToMemoryRatio == 0.0 && config.CPURequestToLimitRatio == 0.0 && config.MemoryRequestToLimitRatio == 0.0 {
+			return config, fmt.Errorf("PodLimitRequest plugin enabled but no ratios specified")
+		}
+		if config.LimitCPUToMemoryRatio < 0.0 {
+			return config, fmt.Errorf("LimitCPUToMemoryRatio must be positive")
+		}
+		if config.CPURequestToLimitRatio < 0.0 || config.CPURequestToLimitRatio > 1.0 {
+			return config, fmt.Errorf("CPURequestToLimitRatio must be between 0.0 and 1.0")
+		}
+		if config.MemoryRequestToLimitRatio < 0.0 || config.MemoryRequestToLimitRatio > 1.0 {
+			return config, fmt.Errorf("MemoryRequestToLimitRatio must be between 0.0 and 1.0")
+		}
+	}
+	return config, err
+}
+
+func (a *podLimitRequest) Admit(attr admission.Attributes) error {
+	glog.V(8).Infof("PodLimitRequest admission controller is invoked")
+	if !a.Config.Enabled {
+		return nil
+	}
+	if pod, isPod := attr.GetObject().(*kapi.Pod); isPod && attr.GetSubresource() == "" {
```

Comment:
I can think it'd be possible to have an API endpoint (say a virtual resource) that takes a pod object but doesn't mean it's creating a pod. I think checking the resource name instead of the object type narrows the scope of what you're accepting to "a pod getting created". @deads2k do you disagree?